test_runner: add global setup and teardown functionality

pmarchini opened the pull request:
I'm opening this PR as a draft to gather some feedback regarding global setup/teardown.

The idea is to leverage the new configuration file to allow users to manage this scenario.

Personally, I'm currently using a "wrapper" file that directly runs the `run` function, preceded and followed by `setup/teardown` when it's necessary to launch containers or other resources.

At the moment, I've added support for a possible shared `context` object between the two hooks.

However, I'm not fully convinced about this approach.

I think some crucial improvements might include adding events into the test stream to notify about the success or failure of global setup and teardown.

Note: this PR is still missing `run` new option documentation and test to cover the feature via `run` function.

<!--
Before submitting a pull request, please read:

- the CONTRIBUTING guide at https://github.com/nodejs/node/blob/HEAD/CONTRIBUTING.md
- the commit message formatting guidelines at
  https://github.com/nodejs/node/blob/HEAD/doc/contributing/pull-requests.md#commit-message-guidelines

For code changes:
1. Include tests for any bug fixes or new features.
2. Update documentation if relevant.
3. Ensure that `make -j4 test` (UNIX), or `vcbuild test` (Windows) passes.

If you believe this PR should be highlighted in the Node.js CHANGELOG
please add the `notable-change` label.

Developer's Certificate of Origin 1.1

By making a contribution to this project, I certify that:

(a) The contribution was created in whole or in part by me and I
    have the right to submit it under the open source license
    indicated in the file; or

(b) The contribution is based upon previous work that, to the best
    of my knowledge, is covered under an appropriate open source
    license and I have the right under that license to submit that
    work with modifications, whether created in whole or in part
    by me, under the same open source license (unless I am
    permitted to submit under a different license), as indicated
    in the file; or

(c) The contribution was provided directly to me by some other
    person who certified (a), (b) or (c) and I have not modified
    it.

(d) I understand and agree that this project and the contribution
    are public and that a record of the contribution (including all
    personal information I submit with it, including my sign-off) is
    maintained indefinitely and may be redistributed consistent with
    this project or the open source license(s) involved.
-->

pmarchini commented:
Although it's still a draft, would anyone be willing to spend some time providing feedback / ideas? @nodejs/test_runner 

Thaaaanks 🚀

cjihrig commented:
I looked at this very quickly, so I may have missed something. There are some additional things that I think need to be tested:

- Running without `--test`. If someone wants to run a single file that contains tests like `node foo.js`, that should work like the other test runner flags do.
- This should verify that no user code can run before these new hooks. This is especially important for `describe()`, which builds the suites before the test runner starts executing things.
- How does this interact with `--import` and `--require` since this is duplicating functionality that they provide.

pmarchini commented:
Hey @cjihrig, thanks for the feedback!

Point by point:  

1.  
I think it would make sense to limit the functionality to `--test` only or direct `run` invocation.  

2.  
I agree, I will cover that case.  

3.  
Currently, `--require` and `--import` clearly act at startup, and in the case of `isolation process`, they cause a rerun for each individual test file.  
I wonder if, from a developer experience perspective, it would make sense to evaluate whether overlapping multiple flags could lead to unintended side effects for the user.  
At the same time, having potential overlap seems reasonable to me...what do you think?  

4.  
Regarding `watch`, I'm torn. Initially, I was thinking of not supporting it, but reconsidering, especially in scenarios involving container creation/teardown or similar cases, it seems important to provide this functionality.  
My only concern is whether it would be possible to break down the implementation incrementally (also to ease the review process).

cjihrig commented:
> I think it would make sense to limit the functionality to --test only or direct run invocation.

> Regarding watch, I'm torn.

I feel pretty strongly that these need to be supported. For people working on the test runner, I understand that it can be annoying to support the various modes of operation, but that isn't a good reason not to do it. IMO, it only makes sense to omit something if there is a good reason from the end user's perspective (for example `--test-force-exit` not being compatible with watch mode).

Regarding interop with `--require`/`--import`, I don't really have strong feelings, but I think we should nail down what the behavior will be. If those always run before this new functionality, that's totally fine. I think we should have a test for the happy path. We probably should test the case where someone uses one of these flags and invokes the test runner from there while also having a global setup.

pmarchini commented:
Sounds good to me, I see your reasons and you obviously have more vision and experience than me so I'm more than glad to follow your suggestions! 🚀
I'll update the PR ASAP

atlowChemi commented on doc/api/cli.md:
````diff
@@ -2464,6 +2464,52 @@ added:
+* A `globalTeardown` function which runs once after all tests complete
+* Or both functions
+
+```js
+// setup-module.js
+async function globalSetup() {
+  // Setup shared resources, state, or environment
+  console.log('Global setup executed');
+  // Run servers, create files, prepare databases, etc.
+}
+
+async function globalTeardown() {
+  // Clean up resources, state, or environment
+  console.log('Global teardown executed');
+  // Close servers, remove files, disconnect from databases, etc.
+}
+
+module.exports = { globalSetup, globalTeardown };
+```
+
+If the module is an ESM module, the functions should be exported using named exports:
+
+```mjs
+// setup-module.mjs
+export async function globalSetup() {
+  // Setup code here
+}
+
+export async function globalTeardown() {
+  // Teardown code here
+}
+```
````
Why not: (the docs would show it as one code block with a toggle of cjs/mjs)
````suggestion
```cjs
// setup-module.js
async function globalSetup() {
  // Setup shared resources, state, or environment
  console.log('Global setup executed');
  // Run servers, create files, prepare databases, etc.
}

async function globalTeardown() {
  // Clean up resources, state, or environment
  console.log('Global teardown executed');
  // Close servers, remove files, disconnect from databases, etc.
}

module.exports = { globalSetup, globalTeardown };
```

```mjs
// setup-module.mjs
export async function globalSetup() {
  // Setup shared resources, state, or environment
  console.log('Global setup executed');
  // Run servers, create files, prepare databases, etc.
}

export async function globalTeardown() {
  // Clean up resources, state, or environment
  console.log('Global teardown executed');
  // Close servers, remove files, disconnect from databases, etc.
}
```
````

pmarchini replied:
I didn't know about this "trick", thanks for the advice!!!

pmarchini replied:
done!

pmarchini replied:
@atlowChemi could I please ask you a quick review?

atlowChemi replied:
@pmarchini I think it still requires some change.
One section should have `cjs` while the other has `mjs`, and the content should be identical, other than the different import/export systems.
AFAICT - that is not the current state

pmarchini replied:
it should be ok now!

atlowChemi commented on src/node_options.cc:
```diff
@@ -761,6 +761,10 @@ EnvironmentOptionsParser::EnvironmentOptionsParser() {
             &EnvironmentOptions::coverage_exclude_pattern,
             kAllowedInEnvvar);
+  AddOption("--test-global-setup", 
+            "run this script before running tests",
```
This doesn't feel accurate IMHO, as it could run a script after running the tests as well

atlowChemi replied:
Also, this should be added to `node.1` as well
https://github.com/nodejs/node/blob/fab6906c5d1c16b046187485a7cb136b3659d42b/doc/node.1

pmarchini replied:
> This doesn't feel accurate IMHO, as it could run a script after running the tests as well 
I totally agree, this needs a better description.
What do you think about the flag name? should we coming up with a better naming ? 

Should we follow mocha's "global-fixtures" ? https://mochajs.org/#global-fixtures ?

atlowChemi replied:
In `jest` and `vitest` it is called `globalSetup`, so IMHO that is a good name, but I don't feel strongly about it

pmarchini replied:
I've seen `jest globalTeardown`, IMHO, having a single file could improve devEx!

atlowChemi replied:
> IMHO, having a single file could improve devEx!

I agree 🙂 
That is also how vitest works

pmarchini commented:
hey @cjihrig, @atlowChemi 

I've added the following tests :
- Basic setup and teardown functionality
- Setup-only modules
- Teardown-only modules
- Multiple test files with shared setup/teardown
- Context sharing between setup and teardown
- Async setup and teardown
- Error handling in setup
- TypeScript support for setup/teardown functions
- ESM module support for setup/teardown functions
- Ensuring setup runs only once per test run
- Ensuring teardown runs only once per test run
- Interop with --require flag (required module before globalSetup)
- Interop with --import flag (behavior differs between run modes)
- Tests in imported modules with setup/teardown
- Watch mode interaction with global setup/teardown
- Direct invocation vs. --test flag behavior
- Edge cases when combining multiple flags

My main concern at the moment is this change in behaviour:
https://github.com/nodejs/node/pull/57438/files#diff-196bc4ec67eaf2b311f953b5282702f42d7c55a5e6ab5599fdf83f51fdf3c23cR197-R210
```js
  it('should include test type in enqueue, dequeue events', async (t) => {
    const stream = await run({
      files: [join(testFixtures, 'default-behavior/test/suite_and_test.cjs')],
    });
    t.plan(4);
    stream.on('test:enqueue', common.mustCall((data) => {
      if (data.name === 'this is a suite') {
        t.assert.strictEqual(data.type, 'suite');
      }
      if (data.name === 'this is a test') {
        t.assert.strictEqual(data.type, 'test');
      }
    }, 3)); // TODO(pmarchini): this should be 2 but after `await root.harness.waitForGlobalSetup()` it's 3
```

I'd really appreciate any feedback or suggestions regarding how I've implemented the feature 😁

marco-ippolito commented on doc/api/cli.md:
```diff
@@ -2464,6 +2464,54 @@ added:
+added: REPLACEME
+-->
+
+> Stability: 1 - Experimental
```
```suggestion
> Stability: 1.0 - Early development
```

https://nodejs.org/api/documentation.html

marco-ippolito commented on doc/api/cli.md:
```diff
@@ -2464,6 +2464,54 @@ added:
+
+> Stability: 1 - Experimental
+
+Specify a module that will be run before all tests are executed and
```
```suggestion
Specify a module that will be evaluated before all tests are executed and
```

I mean the `globalTeardown` will be run after, so I'm not sure (I'm also not sure evaluated is correct)

marco-ippolito commented on doc/api/cli.md:
```diff
@@ -2464,6 +2464,54 @@ added:
+
+Specify a module that will be run before all tests are executed and
+can be used to setup global state or fixtures for tests.
+This module should export either:
+
+* A `globalSetup` function which runs once before all tests start
+* A `globalTeardown` function which runs once after all tests complete
+* Or both functions
```
```suggestion
This module can export:

* A `globalSetup` function which runs once before all tests start
* A `globalTeardown` function which runs once after all tests complete
```

marco-ippolito commented on lib/internal/test_runner/harness.js:
```diff
@@ -29,25 +29,57 @@ const {
 
 testResources.set(reporterScope.asyncId(), reporterScope);
 
+async function setupGlobalSetupTeardownFunctions(globalSetupPath, cwd) {
```
jsdoc would be nice

marco-ippolito commented on lib/internal/test_runner/harness.js:
```diff
@@ -71,16 +103,45 @@ function createTestTree(rootTestOptions, globalOptions) {
+        return;
+      }
+
+      const globalSetupFunctions =
```
```suggestion
      const globalSetupFunctions =
```
This kind of formatting looks weird

pmarchini commented:
Looking after the failing test!

cjihrig commented on lib/internal/test_runner/utils.js:
```diff
@@ -270,6 +272,7 @@ function parseCommandLine() {
         total: NumberParseInt(indexAndTotal[1], 10),
       };
     }
+
```
Can you please remove this.

cjihrig commented on lib/internal/test_runner/runner.js:
```diff
@@ -795,18 +802,25 @@ function run(options = kEmptyObject) {
     if (typeof setup === 'function') {
       await setup(root.reporter);
     }
-
```
Can you please undo this.

cjihrig commented on doc/node.1:
```diff
@@ -464,6 +464,9 @@ Require a minimum threshold for line coverage (0 - 100).
 finished executing even if the event loop would otherwise remain active.
 .
+.It Fl -test-global-setup
+Specify a module that will be run before all tests are executed and can be used to setup global state or fixtures for tests.
```
```suggestion
Specifies a module containing global setup and teardown functions for the test runner.
```

cjihrig commented on lib/internal/test_runner/harness.js:
```diff
@@ -29,25 +29,57 @@ const {
+const esmLoader = require('internal/modules/esm/loader');
+const { kEmptyObject } = require('internal/util');
+const { validateFunction } = require('internal/validators');
+const { getOptionValue } = require('internal/options');
```
We shouldn't be using this function in this file. All of the `getOptionValue()` calls should have already happened in `parseCommandLine()`, and the value should already be available here.

cjihrig commented on lib/internal/test_runner/harness.js:
```diff
@@ -29,25 +29,57 @@ const {
 
 testResources.set(reporterScope.asyncId(), reporterScope);
 
+async function setupGlobalSetupTeardownFunctions(globalSetupPath, cwd) {
```
All of this might fit better in `utils.js`, and called from `parseCommandLine()`, similar to how some other flags like `--test-name-pattern` are processed.

pmarchini replied:
Regarding having it as part of `utils.js`: I agree, it definitely makes more sense.  
As for `parseCommandLine`: at the moment, it is not "`cwd`-aware", and I'm not sure whether it should have that dependency.

mcollina reviewed (approved):
lgtm

cjihrig commented on doc/api/test.md:
```diff
@@ -397,6 +397,60 @@ their dependencies. When a change is detected, the test runner will
+can be used to setup global state or fixtures for tests. This is useful for preparing resources or setting up
+shared state that is required by multiple tests.
+
+This module should export either:
```
"either" seems like the user can only specify one or the other.

cjihrig commented on test/fixtures/test-runner/global-setup-teardown/basic-setup-teardown.ts:
```diff
@@ -0,0 +1,35 @@
+  }
+  
+  // Clean up the setup file too
+  if (setupFlagPath && fs.existsSync(setupFlagPath)) {
```
Please don't use `existsSync()` to check for a file before attempting to do something with it. This is a TOCTOU race condition. Either delete the file and handle the error, or use something like `fs.rm()` with the `force` option.

There are some other uses of this pattern that should be updated as well.

cjihrig commented on test/fixtures/test-runner/global-setup-teardown/basic-setup-teardown.js:
```diff
@@ -0,0 +1,26 @@
+'use strict';
+
+const fs = require('fs');
```
Here and in other places:

```suggestion
const fs = require('node:fs');
```

cjihrig commented on test/fixtures/test-runner/global-setup-teardown/async-setup-teardown.js:
```diff
@@ -0,0 +1,28 @@
+  console.log('Async teardown starting');
+
+  // Simulate an async operation
+  await new Promise((resolve) => setTimeout(resolve, 100));
```
Please use `setTimeout()` from `node:timers/promises` for this type of thing. Here and in other places.

cjihrig commented on lib/internal/test_runner/harness.js:
```diff
@@ -285,6 +325,11 @@ function lazyBootstrapRoot() {
 
 async function startSubtestAfterBootstrap(subtest) {
   if (subtest.root.harness.buildPromise) {
+    if (subtest.root.harness.globalSetupPromise) {
```
This also feels off to me. I would expect the global setup to be part of the bootstrap process since it happens before any tests can run.

pmarchini commented on lib/internal/test_runner/harness.js:
```diff
@@ -94,7 +126,7 @@ function createTestTree(rootTestOptions, globalOptions) {
 function createProcessEventHandler(eventName, rootTest) {
   return (err) => {
-    if (rootTest.harness.bootstrapPromise) {
+    if (rootTest.harness.bootstrapPromise || rootTest.harness.globalSetupPromise) {
```
I agree, my biggest concern at the moment is that we're manually setting `harness.bootstrapPromise` to different values based on the type of run (isolation/watch/lazy bootstrap), and for this reason, I have the impression that it makes it more complex to understand and maintain the code.  

https://github.com/nodejs/node/blob/657f81853253a16ed9557d67d17372498a419a9f/lib/internal/test_runner/runner.js#L725  
https://github.com/nodejs/node/blob/657f81853253a16ed9557d67d17372498a419a9f/lib/internal/test_runner/runner.js#L738  
https://github.com/nodejs/node/blob/657f81853253a16ed9557d67d17372498a419a9f/lib/internal/test_runner/runner.js#L753  
https://github.com/nodejs/node/blob/657f81853253a16ed9557d67d17372498a419a9f/lib/internal/test_runner/harness.js#L281  
https://github.com/nodejs/node/blob/657f81853253a16ed9557d67d17372498a419a9f/lib/internal/test_runner/harness.js#L290 

Having both bootstrap and globalSetup allows, from a code perspective, the decoupling of the existing `bootstrapPromise` logic (shared across the different run modes), with the goal of reducing it to just two distinct flows where the global setup is handled: lazyBootstrap and `run` bootstrap.  

This is the rationale behind the implementation.

cjihrig replied:
I don't think there is anything wrong with having a separate promise. I was referring to the lifetime of the promises. If I am understanding the changes, the global setup is considered part of the bootstrapping process. So, the global setup should always be finished by the time the bootstrapping is considered finished (if it isn't, then things are likely to go wrong). I don't want to get to a state where we have a bunch of different promises that we need to check all over the place.

pmarchini replied:
I've just pushed a refactor to take advantage of `rootTest.harness.bootstrapPromise` to encapsulate the global setup while reducing the feature's footprint.

WDYT?

cjihrig reviewed (approved):
LGTM, thanks!

mcollina reviewed (approved):
lgtm

aduh95 commented:
@nodejs-github-bot test https://github.com/nodejs/node/pull/57438/commits/6981e02441c4d30bc2fa6497f98e3c2b560dba75

cjihrig commented on lib/internal/test_runner/harness.js:
```diff
@@ -250,6 +280,7 @@ function setupProcessState(root, globalOptions) {
   process.on('uncaughtException', exceptionHandler);
   process.on('unhandledRejection', rejectionHandler);
   process.on('beforeExit', exitHandler);
+  process.on('exit', terminationHandler);
```
The event handler being added does not match the one being removed. Why does this need to be added at all though?

pmarchini replied:
Thanks for the catch!  
While debugging the failing tests on Windows using a VM, I noticed differences between the VS2019 and VS2022 builds.  
In the latest version (VS2022), SIGTERM and SIGINT are not handled properly. Sending `process.kill('SIGTERM')` or `process.kill('SIGINT')` to a spawned process wasn’t intercepted, and the teardown wasn’t executed when exiting from a process running with `--watch`.

In the test `test/parallel/test-runner-global-setup-watch-mode.mjs`, replacing `process.exit(0)` with `process.kill('SIGTERM')` or `process.kill('SIGINT')` causes the test to fail only on Windows with VS2022.

cjihrig replied:
I dunno, the details of that bug seem sketchy, and I it seems weird that adding an `'exit'` handler in the test runner is the right way to fix it.

pmarchini replied:
I agree, so far that potentially wasn’t an issue since we didn’t have anything specific in the teardown.  
Now, having a user-defined function requires guaranteeing that the function is actually run.

On the other hand, the very strange thing to me is that this behavior only occurs on that specific Windows build.  
On other systems, it works as expected.  
I ran the CI many times before checking on the VM.

Personally, I don’t see any particular issues with having the `terminationHandler` executed on exit.  
To me, it feels like a final safety net to ensure the teardown is properly executed.

I’m waiting for the CI run just to be sure that this approach really works.

But just to be clear: I don’t like this approach, and my target is to find a better solution.

I could try to replicate the same problem in a separate PR just to better define and isolate the behaviour I found (assuming it’s not due to my changes).

cjihrig replied:
Let's see what the CI says, but to me, having both a `beforeExit` and `exit` handler doing essentially the same thing seems like a smell.

pmarchini replied:
BTW, based on what I've seen, the issue is that a process spawned with `const child = child_process.spawn()` on a Windows Visual Studio 2022 build doesn't handle a `SIGINT` signal when we call `child.kill('SIGINT')`.  

I'll test this in isolation. 
I might be totally wrong.  


Here's an example of a Windows run where `SIGINT` was not handled: https://ci.nodejs.org/job/node-test-binary-windows-js-suites/33492/

pmarchini replied:
<img width="702" alt="image" src="https://github.com/user-attachments/assets/00682008-1bfb-4e30-9248-2362bd3776ad" />
In this case, the tests passed.

This whole `on('exit')` thing is only needed to be able to programmatically stop the test watcher in this specific scenario. However, it's just a workaround.  
I opened another PR just to run the CI and check whether the same issue occurs in isolation:  https://github.com/nodejs/node/pull/57794

pmarchini commented on test/fixtures/test-runner/global-setup-teardown/context-sharing.js:
```diff
@@ -0,0 +1,25 @@
+'use strict';
```
this fixture is no more needed

richardlau commented:
This doesn't land cleanly on v22.x-staging so a manual backport will be necessary if this is to be released there.
